### PR TITLE
Swap out arrange for PyNest2d

### PR DESCRIPTION
**Pull request by nallath**
CURA-7440

Also needs:
https://github.com/Ultimaker/Uranium/pull/643

### CODE REVIEW

File: cura/Arranging/ArrangeObjectsJob.py [hunk 1]
```diff
@@ -1,23 +1,16 @@
...
-i18n_catalog = i18nCatalog("cura")
+from cura.Arranging.Nest2DArrange import arrange
 
-from cura.Scene.ZOffsetDecorator import ZOffsetDecorator
-from cura.Arranging.Arrange import Arrange
-from cura.Arranging.ShapeArray import ShapeArray
+i18n_catalog = i18nCatalog("cura")
 
 from typing import List
+from pynest2d import *
```
Comment (konskarm) [inline cura/Arranging/ArrangeObjectsJob.py:13]: Unused import?

File: cura/Arranging/ArrangeObjectsJob.py [hunk 2]
```diff
@@ -34,76 +27,12 @@ def run(self):
...
 
         status_message.hide()
-
         if not found_solution_for_all:
             no_full_solution_message = Message(i18n_catalog.i18nc("@info:status", "Unable to find a location within the build volume for all objects"),
                                                title = i18n_catalog.i18nc("@info:title", "Can't Find Location"))
             no_full_solution_message.show()
-
-        self.finished.emit(self)
+        self.finished.emit(self)
```
Comment (konskarm) [inline cura/Arranging/ArrangeObjectsJob.py:38]: Add new line at end of file
```suggestion
        self.finished.emit(self)
        
```

File: cura/Arranging/Nest2DArrange.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+def findNodePlacement(nodes_to_arrange: List["SceneNode"], build_volume: "BuildVolume", fixed_nodes: Optional[List["SceneNode"]] = None, factor = 10000) -> Tuple[bool, List[Item]]:
+    """
+    Find placement for a set of scene nodes, but don't actually move them just yet.
+    :param nodes_to_arrange: The list of nodes that need to be moved.
+    :param build_volume: The build volume that we want to place the nodes in. It gets size & disallowed areas from this.
+    :param fixed_nodes: List of nods that should not be moved, but should be used when deciding where the others nodes
+                        are placed.
+    :param factor: The library that we use is int based. This factor defines how accuracte we want it to be.
+    :return:
```
Comment (konskarm) [inline cura/Arranging/Nest2DArrange.py:27]: Missing return documentation

File: cura/Arranging/Nest2DArrange.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+
+def findNodePlacement(nodes_to_arrange: List["SceneNode"], build_volume: "BuildVolume", fixed_nodes: Optional[List["SceneNode"]] = None, factor = 10000) -> Tuple[bool, List[Item]]:
+    """
+    Find placement for a set of scene nodes, but don't actually move them just yet.
+    :param nodes_to_arrange: The list of nodes that need to be moved.
+    :param build_volume: The build volume that we want to place the nodes in. It gets size & disallowed areas from this.
+    :param fixed_nodes: List of nods that should not be moved, but should be used when deciding where the others nodes
+                        are placed.
+    :param factor: The library that we use is int based. This factor defines how accuracte we want it to be.
```
Comment (konskarm) [inline cura/Arranging/Nest2DArrange.py:26]: Typo
```suggestion
    :param factor: The library that we use is int based. This factor defines how accurate we want it to be.
```

File: cura/Arranging/Nest2DArrange.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        num_disallowed_areas_added += 1
+
+    for node in fixed_nodes:
+        converted_points = []
+        hull_polygon = node.callDecoration("getConvexHull")
+        
+        for point in hull_polygon.getPoints():
+            converted_points.append(Point(point[0] * factor, point[1] * factor))
+        item = Item(converted_points)
+        node_items.append(item)
```
Comment (konskarm) [inline cura/Arranging/Nest2DArrange.py:80]: Seems like the same item is added twice in the node_items list (lines 80 and 82).

File: cura/Arranging/Nest2DArrange.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+def arrange(nodes_to_arrange: List["SceneNode"], build_volume: "BuildVolume", fixed_nodes: Optional[List["SceneNode"]] = None, factor = 10000) -> bool:
+    """
+    Find placement for a set of scene nodes, and move them by using a single grouped operation.
+    :param nodes_to_arrange: The list of nodes that need to be moved.
+    :param build_volume: The build volume that we want to place the nodes in. It gets size & disallowed areas from this.
+    :param fixed_nodes: List of nods that should not be moved, but should be used when deciding where the others nodes
+                        are placed.
+    :param factor: The library that we use is int based. This factor defines how accuracte we want it to be.
+    :return:
```
Comment (konskarm) [inline cura/Arranging/Nest2DArrange.py:107]: Missing return documentation

File: cura/MultiplyObjectsJob.py [hunk 7]
```diff
@@ -93,17 +73,28 @@ def run(self) -> None:
...
                 nodes.append(new_node)
-                current_progress += 1
-                status_message.setProgress((current_progress / total_progress) * 100)
-                QCoreApplication.processEvents()
-                Job.yieldThread()
-            QCoreApplication.processEvents()
-            Job.yieldThread()
-
+        factor = 10000
+        found_solution_for_all, node_items = findNodePlacement(nodes, Application.getInstance().getBuildVolume(), fixed_nodes, factor = 10000)
```
Comment (konskarm) [inline cura/MultiplyObjectsJob.py:77]:
```suggestion
        found_solution_for_all, node_items = findNodePlacement(nodes, Application.getInstance().getBuildVolume(), fixed_nodes, factor = factor)
```

File: cura/MultiplyObjectsJob.py [hunk 8]
```diff
@@ -93,17 +73,28 @@ def run(self) -> None:
...
+                    operation.addOperation(RotateOperation(new_node, Quaternion.fromMatrix(rotation_matrix)))
+                    operation.addOperation(
+                        TranslateOperation(new_node, Vector(node_item.translation().x() / factor, 0,
+                                                        node_item.translation().y() / factor)))
+                else:
+                    # We didn't find a spot
+                    operation.addOperation(
+                        TranslateOperation(new_node, Vector(200, 0, -not_fit_count * 20), set_position=True))
+                    not_fit_count += 1
+
```
Comment (konskarm) [inline cura/MultiplyObjectsJob.py:97]: Call Libnest2dArrange.arrange instead of lines 77-97 in MultiplyObjectsJob.py as these lines seems to be doing the exact same thing

File: cura_app.py [hunk 9]
```diff
@@ -22,6 +22,7 @@
 # tries to create PyQt objects on a non-main thread.
 import Arcus  # @UnusedImport
 import Savitar  # @UnusedImport
+import pynest2d # @UnusedImport
```
Comment (konskarm) [inline cura_app.py:25]: Why is this needed?

**Review by konskarm — CHANGES_REQUESTED** [review]

File: cura/MultiplyObjectsJob.py [hunk 10]
```diff
@@ -93,17 +73,28 @@ def run(self) -> None:
...
+                    rotation_matrix = Matrix()
+                    rotation_matrix.setByRotationAxis(node_item.rotation(), Vector(0, -1, 0))
+                    operation.addOperation(RotateOperation(new_node, Quaternion.fromMatrix(rotation_matrix)))
+                    operation.addOperation(
+                        TranslateOperation(new_node, Vector(node_item.translation().x() / factor, 0,
+                                                        node_item.translation().y() / factor)))
+                else:
+                    # We didn't find a spot
+                    operation.addOperation(
+                        TranslateOperation(new_node, Vector(200, 0, -not_fit_count * 20), set_position=True))
```
Comment (konskarm) [inline cura/MultiplyObjectsJob.py:95]: Maybe document a little more why "200" and "-not_fit_count * 20"

File: cura_app.py (hunk 9 above)
Comment (nallath) [reply cura_app.py:25]: For the same reason as the others are there, we get some weird sip pointer error if we don't.

**Comment by rburema** [comment]
CI will need to be updated.
